proof-systems/src/darlin/fiat-shamir/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+
+/// An implementation using Poseidon fs_rng.
+pub mod poseidon;
+
+/// Traits definition for circuitizing FiatShamirRng.
+pub mod constraints;
+
+/// Error types for FiatShamir
+pub mod error;
+
+#[macro_use]
+extern crate derivative;
+
+/// A trait for serialization of [`FiatShamirRng`] seed material.
+#[derive(Default)]
+/// Encoding of seed material as discussed in [issue/22](https://github.com/HorizenLabs/poly-commit/issues/22).
+/// Output type of the seed is a byte array.
+pub struct FiatShamirRngSeed {
+    // the number of seed elements.
+    num_elements: u64,
+    // the byte lengths of the seed elements.
+    elements_len: Vec<u64>,
+    // the concatenated byte sequence of elements.
+    seed_bytes: Vec<u8>,
+}
+
+impl FiatShamirRngSeed {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add_bytes<'a, T: 'a + CanonicalSerialize>(
+        &mut self,
+        elem: &'a T,
+    ) -> Result<&mut Self, Error> {
+        // Check we have not reached the maximum allowed seed size
+        if self.num_elements == u64::MAX {
+            return Err(Error::BadFiatShamirInitialization(format!(
+                "Maximum seed length {} exceeded",
+                u64::MAX
+            )));
+        }
+
+        // Get elem bytes and check that they are not over the maximum allowed elem len
+        let mut elem_bytes = serialize_no_metadata!(elem).map_err(|_| {
+            Error::BadFiatShamirInitialization("Unable to convert elem to bytes".to_owned())
+        })?;
+        let elem_bytes_len: u64 = elem_bytes.len().try_into().map_err(|_| {
+            Error::BadFiatShamirInitialization(format!(
+                "Max elem length exceeded. Max: {}",
+                u64::MAX
+            ))
+        })?;
+
+        // Update internal state
+        self.num_elements += 1;
+        self.elements_len.push(elem_bytes_len);
+        self.seed_bytes.append(&mut elem_bytes);
+        Ok(self)
+    }
+
+    pub fn finalize(self) -> Result<Vec<u8>, Error> {
+        serialize_no_metadata![self.num_elements, self.elements_len, self.seed_bytes].map_err(|e| {
+            Error::BadFiatShamirInitialization(format!("Unable to finalize seed: {:?}", e))
+        })
+    }
+}
+
+/// Thin wrapper around a type which is ToConstraintField + CanonicalSerialize
+pub trait Recordable<F: Field>: ToConstraintField<F> + CanonicalSerialize {}
+
+impl<F: Field, T: ToConstraintField<F> + CanonicalSerialize> Recordable<F> for T {}
+
+/// General trait for Fiat-Shamir transform, designed as a fs_rng-based construction.
+pub trait FiatShamirRng: Sized + Default {
+    /// Internal State
+    type State: Clone + Debug + Eq + PartialEq;
+
+    /// Create a new `Self` by initializing its internal state with a fresh `seed`.
+    fn from_seed(seed: Vec<u8>) -> Result<Self, Error>;
+
+    /// Record new data as part of this Fiat-Shamir transform
+    fn record<F: Field, R: Recordable<F>>(&mut self, data: R) -> Result<&mut Self, Error>;
```

Comment:
This signature allows to chain calls to record(), much like it's done for updatable hashes. I.e. you can do:

`fs_rng.record(a).record(b).record(c);`

Instead of:

```
fs_rng.record(a);
fs_rng.record(b);
fs_rng.record(c);
```

Reply:
Makes sense, marking it as solved